% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/DIALOGUE.main.R
\name{DIALOGUE.run}
\alias{DIALOGUE.run}
\title{DIALOGUE.run}
\usage{
DIALOGUE.run(rA, config = dialogue_config(F))
}
\arguments{
\item{rA}{list of \linkS4class{cell.type} objects.}

\item{k}{the number of multicellular programs to identify;}

\item{results.dir}{path to the results directory, where the output will be saved;}

\item{plot.flag}{if TRUE then \code{\link{DIALOGUE.plot}} will be called to plot the results; default is FALSE;}

\item{abn.c}{the minimal number of cells that a sample should have to be considered for MCP detection;}

\item{spatial.flag}{should be TRUE if working with spatial data with small niches, and TRUE if working with single cell data or larger tissue microenvironment niches. The default value is FALSE.}

\item{add.effects}{whether to add additional covariates to the multilevel model;
if TRUE then the covariates will extracted from the [conf] slot in the \linkS4class{cell.type} objects;
default is FALSE.}
}
\value{
A list with the following components:

sig -  the multicellular programs, given as a list of signatures;

scores - the multicellular programs' scores in each cell;

gene.pval - the cross-cell-type p-values of each program;

pref - the correlation (R) and association (mixed-effects p-value) between the cell-type-sepcific
components of each multicellular programs.
}
\description{
DIALOGUE is a dimensionality reduction approach that uses cross-cell-type
associations to identify multicellular programs and map the cell transcriptome
as a function of its environment.
}
\seealso{
See \href{https://github.com/livnatje/DIALOGUE}{DIALOGUE GitHub page} for more details.
\code{\link{DIALOGUE.plot}}
}
\author{
Livnat Jerby-Arnon
}
